File: src/main/java/com/iot/temperature/controllers/TemperatureController.java
package com.iot.temperature.controllers;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.temporal.ChronoUnit;
import java.time.temporal.TemporalAmount;
import java.time.temporal.TemporalUnit;
import java.util.*;
import java.util.concurrent.*;

import com.iot.temperature.model.TempPerHour;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.ModelAndView;

import com.iot.temperature.model.Temperature;
import com.iot.temperature.repository.TemperatureRepository;
import com.iot.temperature.service.TemperatureService;

@RestController
public class TemperatureController {
	private final Logger log = LoggerFactory.getLogger(this.getClass());

	@Autowired
	private TemperatureRepository temperatureRepo;
	
	@Autowired
	private TemperatureService service;

	@PostMapping("/temperature/{temperature}/humidity/{humidity}/pressure/{pressure}")
	public ModelAndView createPayment(@PathVariable Optional<Float> temperature, @PathVariable Optional<Float> humidity, @PathVariable Optional<Float> pressure) {

//		 log.info("temperature: " + temperature);
//		 log.info("humidity: " + humidity);
//		 log.info("pressure: " + pressure);

		Temperature temp = new Temperature();
		temp.setTemperature(temperature.get());
		temp.setHumidity(humidity.get());
		temp.setPressure(pressure.get());
		temp.setTimestamp(LocalDateTime.now());
		temperatureRepo.save(temp);

		return new ModelAndView("redirect:/checkoutReservation/");
	}

	@GetMapping(path = "/temperature/{duration}", produces = MediaType.APPLICATION_JSON_VALUE)
	public ResponseEntity<Object> getTemperatureData(@PathVariable Integer duration) {
		List<Map<String, Number>> allValues = new ArrayList<>();
		Map<LocalDateTime, Double> allAverages = new HashMap<>();

		LocalDateTime now = LocalDateTime.now();
		now = now.withNano(0).withHour(0);

		if(duration == 1) {

			for (int i = 0; i < 30; i++) {
				List<Temperature> collect = service.getTemperaturesForDay(now);

				//TODO this needs to be broken down per hour
				OptionalDouble average = collect.stream().mapToDouble(c -> c.getTemperature()).average();

				if(average.isEmpty()) {
					allAverages.put(now, 0d);
					System.out.println(0);
				} else {
					allAverages.put(now, average.getAsDouble());
					System.out.println(average.getAsDouble());
				}

				now = now.minus(1, ChronoUnit.DAYS);
			}

			System.out.println(allAverages);

//			IntSummaryStatistics stats = new IntSummaryStatistics();
//			stats.accept(1);
//			stats.accept(3);
//			stats.getAverage();

		} else {
			List<Temperature> collect = service.getTemperatures(duration);

			for (Temperature temperature : collect) {
				HashMap<String, Number> map = new HashMap<>();

				map.put("temperature", temperature.getTemperature());
				map.put("time", Date.from( temperature.getTimestamp().atZone( ZoneId.systemDefault()).toInstant()).getTime());
				map.put("humidity", temperature.getHumidity());
				map.put("pressure", temperature.getPressure());
				allValues.add(map);
			}
		}

		return new ResponseEntity<Object>(allValues, HttpStatus.OK);
	}

	@GetMapping(path = "/temperature/runningAverage", produces = MediaType.APPLICATION_JSON_VALUE)
	public ResponseEntity<Object> getTemperatureData() throws InterruptedException {
		ConcurrentSkipListMap<LocalDateTime, Number> hourlyAverages = new ConcurrentSkipListMap<>();
		ExecutorService executor = Executors.newFixedThreadPool(100);
		LocalDateTime today = LocalDateTime.now().withNano(0).withSecond(0).withMinute(0).withHour(0);
		List<Map<String, Number>> allValues = new ArrayList<>();

		for (int i = 0; i < 30; i++) {
			for(int j = 0; j < 23; j++) {
				LocalDateTime finalToday = today;
				executor.execute(() -> doStuff(finalToday, hourlyAverages));
				today = today.plus(1, ChronoUnit.HOURS);
			}

			today = today.minus(1, ChronoUnit.DAYS).withHour(0);;
		}

		while(hourlyAverages.size() != 690) {
			Thread.sleep(100);
		}

		today = LocalDateTime.now().plus(1, ChronoUnit.DAYS).minus(1, ChronoUnit.HOURS).withNano(0).withSecond(0).withMinute(0);

		for (int k = 0; k < 30; k++) {
			for(int i = 0; i < 6; i++) {
				List<Number> averageForFourHours = new ArrayList<>();

				for(int j = 0; j < 4; j++) {
					averageForFourHours.add(hourlyAverages.getOrDefault(today, 0));
					today = today.minus(1, ChronoUnit.HOURS);
				}

//				System.out.println(today.minus(4, ChronoUnit.HOURS) + " " + averageForFourHours.stream().mapToDouble(n -> n.doubleValue()).average());
				Map<String, Number> range = new HashMap<>();
				range.put("temperature", averageForFourHours.stream().mapToDouble(n -> n.doubleValue()).average().getAsDouble());
				range.put("time", Date.from(today.minus(4, ChronoUnit.HOURS).atZone(ZoneId.systemDefault()).toInstant()).getTime());
				System.err.println(averageForFourHours.stream().mapToDouble(n -> n.doubleValue()).average().getAsDouble() + " " + today.minus(4, ChronoUnit.HOURS));
				allValues.add(range);

			}

			today = today.minus(1, ChronoUnit.DAYS).withHour(0);
		}
		return new ResponseEntity<Object>(allValues, HttpStatus.OK);
	}

	private void doStuff(LocalDateTime today, ConcurrentSkipListMap<LocalDateTime, Number> hourlyAverages) {
//		System.out.println("Looking for: " + today + " and : " + today.plus(1, ChronoUnit.HOURS)) ;
		List<Temperature> collect = service.getTemperatures(today, today.plus(1, ChronoUnit.HOURS));
		System.out.println("Looking for: " + today + " and : " + today.plus(1, ChronoUnit.HOURS) + " found: " + collect.size());

		OptionalDouble average = collect.parallelStream().mapToDouble(c -> c.getTemperature()).average();
		hourlyAverages.put(today, average.isEmpty() ? 0 : average.getAsDouble());
	}
}

File: src/main/kotlin/com/iot/temperature/controllers/TemperatureController.kt
package com.iot.temperature.controllers

import com.iot.temperature.model.Temperature
import com.iot.temperature.repository.TemperatureRepository
import com.iot.temperature.service.TemperatureService
import org.jetbrains.annotations.NotNull
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.servlet.ModelAndView
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.temporal.ChronoUnit
import java.util.*
import java.util.concurrent.ConcurrentSkipListMap
import java.util.concurrent.Executors
import javax.cache.Cache

@RestController
class TemperatureController {
    private val log = LoggerFactory.getLogger(this.javaClass)

    @Autowired
    lateinit var temperatureRepo: TemperatureRepository

    @Autowired
    lateinit var service: TemperatureService

    @Autowired
    var cache: Cache<Int, ConcurrentSkipListMap<LocalDateTime, Number>>? = null

    val latestRecord: ResponseEntity<Temperature?>
        @GetMapping(path = ["/temperature/getLatest"], produces = [MediaType.APPLICATION_JSON_VALUE])
        get() {
            val topByTimestampDesc = temperatureRepo!!.findTopByOrderByIdDesc()
            log.info("Latest record: $topByTimestampDesc.timestamp.format(DateTimeFormatter.ISO_DATE_TIME))")
            return ResponseEntity(topByTimestampDesc, HttpStatus.OK)
        }

    @GetMapping(path = ["/temperature/runningAverage/{totalDays}/{hourWindow}"], produces = [MediaType.APPLICATION_JSON_VALUE])
    fun runningAverage(@PathVariable totalDays: Int?, @PathVariable @NotNull hourWindow: Int): ResponseEntity<Any> {
        val allValues = ArrayList<Map<String, Number>>()
        val today = LocalDateTime.now().withNano(0).withSecond(0).withMinute(0).withHour(0)
        val totalHours: Int = totalDays?.times(24) ?: 0
        val hourlyAverages = concurrentSkipListMap(hourWindow, totalHours, today)

        for (i in 0..totalHours.div(hourWindow)) {
            val averageForFourHours = ArrayList<Number>()

            for (j in 1..hourWindow) {
                averageForFourHours.add((hourlyAverages).getOrDefault(today.minusHours(i.times(hourWindow).plus(j).toLong()), 0))
            }
            val range = HashMap<String, Number>()

            range["temperature"] = averageForFourHours.stream().mapToDouble { n -> n.toDouble() }.average().asDouble
            range["time"] = Date.from(today.minusHours((i * hourWindow).toLong()).atZone(ZoneId.systemDefault()).toInstant()).time
            allValues.add(range)
        }
        return ResponseEntity(allValues, HttpStatus.OK)
    }

    private fun concurrentSkipListMap(cacheKey: Int, totalHours: Int, today: LocalDateTime): ConcurrentSkipListMap<LocalDateTime, Number> {
        val executor = Executors.newFixedThreadPool(10)
        var hourlyAverages = ConcurrentSkipListMap<LocalDateTime, Number>()

        if (cache!!.containsKey(cacheKey)) {
            log.info("Found entry in cache")
            hourlyAverages = cache!!.get(cacheKey)
        } else {
            for (i in 0..totalHours) {
                executor.execute { doStuff(today.minus(i.toLong(), ChronoUnit.HOURS), hourlyAverages) }
            }

            while (hourlyAverages.size != totalHours) Thread.sleep(10)
            cache!!.put(cacheKey, hourlyAverages)
        }
        return hourlyAverages
    }


    @PostMapping("/temperature/{temperature}/humidity/{humidity}/pressure/{pressure}")
    fun createPayment(@PathVariable temperature: Optional<Double>, @PathVariable humidity: Optional<Float>, @PathVariable pressure: Optional<Float>): ModelAndView {
        val temp = Temperature()
        temp.temperature = temperature.get()
        temp.humidity = humidity.get()
        temp.pressure = pressure.get()
        temp.timestamp = LocalDateTime.now()
        temperatureRepo!!.save(temp)

        return ModelAndView("redirect:/checkoutReservation/")
    }

    @GetMapping(path = ["/temperature/{duration}"], produces = [MediaType.APPLICATION_JSON_VALUE])
    fun getTemperatureData(@PathVariable duration: Int?): ResponseEntity<Any> {
        val allValues = ArrayList<Map<String, Number>>()
        val allAverages = HashMap<LocalDateTime, Double>()

        var now = LocalDateTime.now().withNano(0).withHour(0)

        if (duration == 1) {
            for (i in 0..29) {
                val collect = service!!.getTemperaturesForDay(now)

                //TODO this needs to be broken down per hour
                val average = collect.stream().mapToDouble { c -> c!!.temperature!!.toDouble() }.average()

                if (average.isEmpty) {
                    allAverages[now] = 0.0
                } else {
                    allAverages[now] = average.asDouble
                }

                now = now.minus(1, ChronoUnit.DAYS)
            }
        } else {
            for (temperature in service!!.getTemperatures(duration!!)) {
                val map = HashMap<String, Number>()

                map["temperature"] = temperature.temperature!!.toDouble()
                map["time"] = Date.from(temperature.timestamp!!.atZone(ZoneId.systemDefault()).toInstant()).time
                map["humidity"] = temperature.humidity!!.toDouble()
                map["pressure"] = temperature.pressure!!.toDouble()

                allValues.add(map)
            }
        }

        return ResponseEntity(allValues, HttpStatus.OK)
    }

    private fun doStuff(today: LocalDateTime, hourlyAverages: ConcurrentSkipListMap<LocalDateTime, Number>) {
        val average = service!!.getTemperatures(today, today.plus(1, ChronoUnit.HOURS))
                .parallelStream()
                .mapToDouble { c -> c!!.temperature!!.toDouble() }
                .average()

        hourlyAverages[today] = average.orElseGet { 0.0 }
//        log.info("Looking for: " + today + " and : " + today.plus(1, ChronoUnit.HOURS) + "  - Average: " + if (average.isEmpty) 0 else average.asDouble)
    }
}